In [1]:
# Some basic Data Libraries
import numpy as np  
import pandas as pd

# All rows and columns display
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

# Ignore warnings
import warnings
warnings.filterwarnings('ignore')

# Cargamos nuestro dataset de Train:

In [2]:
df = pd.read_csv('./data/train.csv', sep=';', decimal='.')
df.head(5).T

0                                  1  \
trans_date_trans_time  2019-02-11 14:48:59                2020-03-22 21:54:16   
cc_num                     213157767990030                   3553629419254918   
category                      shopping_net                          kids_pets   
amt                                   4.39                              14.96   
gender                                   F                                  F   
city                          Rock Springs                             Conway   
state                                   WY                                 WA   
zip                                  82901                              98238   
lat                                 41.606                              48.34   
long                               -109.23                          -122.3456   
city_pop                             27971                                 85   
job                        Music therapist  Research officer, political party   
merch_lat                         41.97153                          49.088724   
merch_long                     -109.512691                        -123.344203   
is_fraud                                 0                                  0   
Year                                  2019                               2020   
Month                                    2                                  3   
state_abbr                              WY                                 WA   
Unemployed percent                     3.6                                5.2   
close_s&p_500                       2709.8                             2237.4   
change_usd_eur                      0.8828                             0.9349   
date                            2019-02-11                         2020-03-22   
crime_rate                           215.0                              293.7   
age                                     39                                 39   
week_number                              7                                 12   
day_of_week                         Monday                             Sunday   
day                                     11                                 22   
time                              14:48:59                           21:54:16   
hour                                    14                                 21   
is_holiday                           False                              False   

                                                        2  \
trans_date_trans_time                 2019-07-07 10:33:08   
cc_num                                    376028110684021   
category                                         misc_pos   
amt                                                 32.43   
gender                                                  M   
city                                            Meadville   
state                                                  MO   
zip                                                 64659   
lat                                               39.7795   
long                                             -93.3014   
city_pop                                              964   
job                    Tourist information centre manager   
merch_lat                                       39.155293   
merch_long                                     -92.421151   
is_fraud                                                0   
Year                                                 2019   
Month                                                   7   
state_abbr                                             MO   
Unemployed percent                                    3.1   
close_s&p_500                                     2975.95   
change_usd_eur                                      0.891   
date                                           2019-07-07   
crime_rate                                          499.6   
age                                                    48   
week

Chequeamos que el shape siga siendo como antes:

In [3]:
df.shape

(1481915, 30)

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1481915 entries, 0 to 1481914
Data columns (total 30 columns):
 #   Column                 Non-Null Count    Dtype  
---  ------                 --------------    -----  
 0   trans_date_trans_time  1481915 non-null  object 
 1   cc_num                 1481915 non-null  int64  
 2   category               1481915 non-null  object 
 3   amt                    1481915 non-null  float64
 4   gender                 1481915 non-null  object 
 5   city                   1481915 non-null  object 
 6   state                  1481915 non-null  object 
 7   zip                    1481915 non-null  int64  
 8   lat                    1481915 non-null  float64
 9   long                   1481915 non-null  float64
 10  city_pop               1481915 non-null  int64  
 11  job                    1481915 non-null  object 
 12  merch_lat              1481915 non-null  float64
 13  merch_long             1481915 non-null  float64
 14  is_fraud          

Vemos que hay una variable repetida, que habíamos usado previamente en la fase de merge con otros datasets: state con state_abbr, dropeamos la segunda:

In [5]:
df = df.drop(['state_abbr'], axis = 1)

In [6]:
df.head()

trans_date_trans_time            cc_num      category     amt gender  \
0   2019-02-11 14:48:59   213157767990030  shopping_net    4.39      F   
1   2020-03-22 21:54:16  3553629419254918     kids_pets   14.96      F   
2   2019-07-07 10:33:08   376028110684021      misc_pos   32.43      M   
3   2020-06-27 23:41:53   180094419304907          home   66.36      F   
4   2019-05-15 12:12:17  3501942333500073          home  115.35      F   

           city state    zip      lat      long  city_pop  \
0  Rock Springs    WY  82901  41.6060 -109.2300     27971   
1        Conway    WA  98238  48.3400 -122.3456        85   
2     Meadville    MO  64659  39.7795  -93.3014       964   
3      Sun City    CA  92585  33.7467 -117.1721     54287   
4       Phoenix    AZ  85086  33.8155 -112.1202   1312922   

                                  job  merch_lat  merch_long  is_fraud  Year  \
0                     Music therapist  41.971530 -109.512691         0  2019   
1   Research officer, political party  49.088724 -123.344203         0  2020   
2  Tourist information centre manager  39.155293  -92.421151         0  2019   
3        Designer, exhibition/display  33.720708 -116.381735         0  2020   
4            Counselling psychologist  34.542705 -111.282381         0  2019   

   Month  Unemployed percent  close_s&p_500  change_usd_eur        date  \
0      2                 3.6        2709.80          0.8828  2019-02-11   
1      3                 5.2        2237.40          0.9349  2020-03-22   
2      7                 3.1        2975.95          0.8910  2019-07-07   
3      6                13.8        3053.24          0.8907  2020-06-27   
4      5                 4.8        2850.96          0.8919  2019-05-15   

   crime_rate  age  week_number day_of_week  day      time  hour  is_holiday  
0       215.0   39            7      Monday   11  14:48:59    14       False  
1       293.7   39           12      Sunday   22  21:54:16    21       False  
2       499.6   48           27      Sunday    7  10:33:08    10       False  
3       442.0   51           26    Saturday   27  23:41:53    23       False  
4       447.1   23           20   Wednesday   15  12:12:17    12       False

## Estudio sobre las variables categóricas

Otras variables que son categóricas: *category*, *city*, *state*, *job* y *state_name*

Vamos a ver cuántas "categorías" diferentes tenemos en cada variable, para hacernos una idea de la dimensionalidad a codificar en valor numérico:

In [7]:
df.select_dtypes(include=['object']).describe()

trans_date_trans_time       category   gender        city    state  \
count                1481915        1481915  1481915     1481915  1481915   
unique               1460810             14        2         906       51   
top      2020-12-13 17:53:47  gas_transport        F  Birmingham       TX   
freq                       4         150564   811662        6444   108167   

                      job        date day_of_week      time  
count             1481915     1481915     1481915   1481915  
unique                497         730           7     86400  
top     Film/video editor  2020-11-30      Monday  23:33:40  
freq                11134        5190      295376        47

Respecto a las fechas, ya habíamos hecho el ejercicio de desagregar `trans_date_trans_time` en día, mes, etc. así que sin problema podemos eliminar `trans_date_trans_time`, `time` y `date`. `day_of_week` tenemos que pasarla a valor numérico

## Codificación de variables numéricas:

1) Codificamos `day_of_week`:

Hacemos el proceso inverso al que hicimos en el Notebook 1:

In [8]:
# map the name of day of the week number to its number
df['day_of_week'] = df['day_of_week'].map({
    'Monday': 0,
    'Tuesday': 1,
    'Wednesday': 2,
    'Thursday': 3,
    'Friday': 4,
    'Saturday': 5,
    'Sunday': 6
})

In [9]:
df.head()

trans_date_trans_time            cc_num      category     amt gender  \
0   2019-02-11 14:48:59   213157767990030  shopping_net    4.39      F   
1   2020-03-22 21:54:16  3553629419254918     kids_pets   14.96      F   
2   2019-07-07 10:33:08   376028110684021      misc_pos   32.43      M   
3   2020-06-27 23:41:53   180094419304907          home   66.36      F   
4   2019-05-15 12:12:17  3501942333500073          home  115.35      F   

           city state    zip      lat      long  city_pop  \
0  Rock Springs    WY  82901  41.6060 -109.2300     27971   
1        Conway    WA  98238  48.3400 -122.3456        85   
2     Meadville    MO  64659  39.7795  -93.3014       964   
3      Sun City    CA  92585  33.7467 -117.1721     54287   
4       Phoenix    AZ  85086  33.8155 -112.1202   1312922   

                                  job  merch_lat  merch_long  is_fraud  Year  \
0                     Music therapist  41.971530 -109.512691         0  2019   
1   Research officer, political party  49.088724 -123.344203         0  2020   
2  Tourist information centre manager  39.155293  -92.421151         0  2019   
3        Designer, exhibition/display  33.720708 -116.381735         0  2020   
4            Counselling psychologist  34.542705 -111.282381         0  2019   

   Month  Unemployed percent  close_s&p_500  change_usd_eur        date  \
0      2                 3.6        2709.80          0.8828  2019-02-11   
1      3                 5.2        2237.40          0.9349  2020-03-22   
2      7                 3.1        2975.95          0.8910  2019-07-07   
3      6                13.8        3053.24          0.8907  2020-06-27   
4      5                 4.8        2850.96          0.8919  2019-05-15   

   crime_rate  age  week_number  day_of_week  day      time  hour  is_holiday  
0       215.0   39            7            0   11  14:48:59    14       False  
1       293.7   39           12            6   22  21:54:16    21       False  
2       499.6   48           27            6    7  10:33:08    10       False  
3       442.0   51           26            5   27  23:41:53    23       False  
4       447.1   23           20            2   15  12:12:17    12       False

2) Eliminamos `trans_date_trans_time` y `time`, porque ya tenemos esa información desagregada y numéricamente:

In [10]:
df = df.drop(['trans_date_trans_time', 'time', 'date'], axis = 1)

In [11]:
df.head()

cc_num      category     amt gender          city state    zip  \
0   213157767990030  shopping_net    4.39      F  Rock Springs    WY  82901   
1  3553629419254918     kids_pets   14.96      F        Conway    WA  98238   
2   376028110684021      misc_pos   32.43      M     Meadville    MO  64659   
3   180094419304907          home   66.36      F      Sun City    CA  92585   
4  3501942333500073          home  115.35      F       Phoenix    AZ  85086   

       lat      long  city_pop                                 job  merch_lat  \
0  41.6060 -109.2300     27971                     Music therapist  41.971530   
1  48.3400 -122.3456        85   Research officer, political party  49.088724   
2  39.7795  -93.3014       964  Tourist information centre manager  39.155293   
3  33.7467 -117.1721     54287        Designer, exhibition/display  33.720708   
4  33.8155 -112.1202   1312922            Counselling psychologist  34.542705   

   merch_long  is_fraud  Year  Month  Unemployed percent  close_s&p_500  \
0 -109.512691         0  2019      2                 3.6        2709.80   
1 -123.344203         0  2020      3                 5.2        2237.40   
2  -92.421151         0  2019      7                 3.1        2975.95   
3 -116.381735         0  2020      6                13.8        3053.24   
4 -111.282381         0  2019      5                 4.8        2850.96   

   change_usd_eur  crime_rate  age  week_number  day_of_week  day  hour  \
0          0.8828       215.0   39            7            0   11    14   
1          0.9349       293.7   39           12            6   22    21   
2          0.8910       499.6   48           27            6    7    10   
3          0.8907       442.0   51           26            5   27    23   
4          0.8919       447.1   23           20            2   15    12   

   is_holiday  
0       False  
1       False  
2       False  
3       False  
4       False

3) Vamos a codificar `gender` con `LabelEncoder`:

In [12]:
from sklearn import preprocessing

le_gender = preprocessing.LabelEncoder()
df["gender"] = le_gender.fit_transform(df["gender"])

4) Para los merchants usaremos one-hot-encoding, ya que tampoco nos va a aumentar mucho la dimensionalidad. Y presuponemos que estas categorías son fijas y vienen dadas, por ejemplo, de una BBDD del Banco/Tarjeta con los tipos de comercios estándar. Aunque, para evitar problemas al evaluar el conjunto de Test o en fase de producción, activamos la opción `handle_unknow = 'ignore` en el `OneHotEncoder`, de forma que cuando el modelo se encuentre una categoría desconocida durante la transformación, las columnas resultantes codificadas en one-hot para esta característica serán todos ceros. En la transformación inversa, una categoría desconocida se denotará como None

Definimos cómo se llamarán las *features* del one-hot con `concat`:

In [13]:
## Define a custom namer:
# def custom_combiner(feature, category):
#     return str(feature) + "_" + type(category).__name__ + "_" + str(category)

In [14]:
ohe_category = preprocessing.OneHotEncoder(feature_name_combiner = 'concat', handle_unknown='ignore')
cat_encoded = ohe_category.fit_transform(df[["category"]])

Va a concatenar la variable `category` con cada uno de los valores:

In [15]:
merch_cats = ohe_category.get_feature_names_out()

Concatenamos las columnas generadas con este One Hot Encoder con las demás del dataframe, eliminando la correspondiente a `category` que ya no nos sirve:

In [16]:
# Concatenate sparse matrix with the Dataframe, droping "old" text-like category column
df = pd.concat([df.drop(columns=["category"]), pd.DataFrame(cat_encoded.toarray(),columns=merch_cats)], axis=1)

In [17]:
df.head()

cc_num     amt  gender          city state    zip      lat  \
0   213157767990030    4.39       0  Rock Springs    WY  82901  41.6060   
1  3553629419254918   14.96       0        Conway    WA  98238  48.3400   
2   376028110684021   32.43       1     Meadville    MO  64659  39.7795   
3   180094419304907   66.36       0      Sun City    CA  92585  33.7467   
4  3501942333500073  115.35       0       Phoenix    AZ  85086  33.8155   

       long  city_pop                                 job  merch_lat  \
0 -109.2300     27971                     Music therapist  41.971530   
1 -122.3456        85   Research officer, political party  49.088724   
2  -93.3014       964  Tourist information centre manager  39.155293   
3 -117.1721     54287        Designer, exhibition/display  33.720708   
4 -112.1202   1312922            Counselling psychologist  34.542705   

   merch_long  is_fraud  Year  Month  Unemployed percent  close_s&p_500  \
0 -109.512691         0  2019      2                 3.6        2709.80   
1 -123.344203         0  2020      3                 5.2        2237.40   
2  -92.421151         0  2019      7                 3.1        2975.95   
3 -116.381735         0  2020      6                13.8        3053.24   
4 -111.282381         0  2019      5                 4.8        2850.96   

   change_usd_eur  crime_rate  age  week_number  day_of_week  day  hour  \
0          0.8828       215.0   39            7            0   11    14   
1          0.9349       293.7   39           12            6   22    21   
2          0.8910       499.6   48           27            6    7    10   
3          0.8907       442.0   51           26            5   27    23   
4          0.8919       447.1   23           20            2   15    12   

   is_holiday  category_entertainment  category_food_dining  \
0       False                     0.0                   0.0   
1       False                     0.0                   0.0   
2       False                     0.0                   0.0   
3       False                     0.0                   0.0   
4       False                     0.0                   0.0   

   category_gas_transport  category_grocery_net  category_grocery_pos  \
0                     0.0                   0.0                   0.0   
1                     0.0                   0.0                   0.0   
2                     0.0                   0.0                   0.0   
3                     0.0                   0.0                   0.0   
4                     0.0                   0.0                   0.0   

   category_health_fitness  category_home  category_kids_pets  \
0                      0.0            0.0                 0.0   
1                      0.0            0.0                 1.0   
2                      0.0            0.0                 0.0   
3                      0.0            1.0                 0.0   
4                      0.0            1.0                 0.0   

   category_misc_net  category_misc_pos  category_personal_care  \
0                0.0                0.0                     0.0   
1                0.0                0.0                     0.0   
2                0.0                1.0                     0.0   
3                0.0                0.0                     0.0   
4                0.0                0.0                     0.0   

   category_shopping_net  category_shopping_pos  category_travel  
0                    1.0                    0.0              0.0  
1                    0.0                    0.0              0.0  
2                    0.0                    0.0              0.0  
3                    0.0                    0.0              0.0  
4                    0.0                    0.0              0.0

5) Para el caso de las profesiones, tenemos muchísimos valores distintos y aquí es muy probable que aparezcan nuevos en el conjunto de Test / en producción. Optamos por elegir un `target encoder`, que establece una relación matemática entre la variable categórica y la variable objetivo:

In [18]:
from sklearn.preprocessing import TargetEncoder

In [19]:
targ_job = preprocessing.TargetEncoder(target_type = 'binary', random_state = 17)

X = df[['job']]
y = df[['is_fraud']]

job_encoded = targ_job.fit_transform(X, y)

In [20]:
te_job = targ_job.feature_names_in_

Concatenamos las columnas generadas con este One Hot Encoder con las demás del dataframe, eliminando la correspondiente a category que ya no nos sirve:

In [21]:
df = pd.concat([df.drop(columns=["job"]), pd.DataFrame(job_encoded, columns = te_job)], axis=1)

In [22]:
df.head()

cc_num     amt  gender          city state    zip      lat  \
0   213157767990030    4.39       0  Rock Springs    WY  82901  41.6060   
1  3553629419254918   14.96       0        Conway    WA  98238  48.3400   
2   376028110684021   32.43       1     Meadville    MO  64659  39.7795   
3   180094419304907   66.36       0      Sun City    CA  92585  33.7467   
4  3501942333500073  115.35       0       Phoenix    AZ  85086  33.8155   

       long  city_pop  merch_lat  merch_long  is_fraud  Year  Month  \
0 -109.2300     27971  41.971530 -109.512691         0  2019      2   
1 -122.3456        85  49.088724 -123.344203         0  2020      3   
2  -93.3014       964  39.155293  -92.421151         0  2019      7   
3 -117.1721     54287  33.720708 -116.381735         0  2020      6   
4 -112.1202   1312922  34.542705 -111.282381         0  2019      5   

   Unemployed percent  close_s&p_500  change_usd_eur  crime_rate  age  \
0                 3.6        2709.80          0.8828       215.0   39   
1                 5.2        2237.40          0.9349       293.7   39   
2                 3.1        2975.95          0.8910       499.6   48   
3                13.8        3053.24          0.8907       442.0   51   
4                 4.8        2850.96          0.8919       447.1   23   

   week_number  day_of_week  day  hour  is_holiday  category_entertainment  \
0            7            0   11    14       False                     0.0   
1           12            6   22    21       False                     0.0   
2           27            6    7    10       False                     0.0   
3           26            5   27    23       False                     0.0   
4           20            2   15    12       False                     0.0   

   category_food_dining  category_gas_transport  category_grocery_net  \
0                   0.0                     0.0                   0.0   
1                   0.0                     0.0                   0.0   
2                   0.0                     0.0                   0.0   
3                   0.0                     0.0                   0.0   
4                   0.0                     0.0                   0.0   

   category_grocery_pos  category_health_fitness  category_home  \
0                   0.0                      0.0            0.0   
1                   0.0                      0.0            0.0   
2                   0.0                      0.0            0.0   
3                   0.0                      0.0            1.0   
4                   0.0                      0.0            1.0   

   category_kids_pets  category_misc_net  category_misc_pos  \
0                 0.0                0.0                0.0   
1                 1.0                0.0                0.0   
2                 0.0                0.0                1.0   
3                 0.0                0.0                0.0   
4                 0.0                0.0                0.0   

   category_personal_care  category_shopping_net  category_shopping_pos  \
0                     0.0                    1.0                    0.0   
1                     0.0                    0.0                    0.0   
2                     0.0                    0.0                    0.0   
3                     0.0                    0.0                    0.0   
4                     0.0                    0.0                    0.0   

   category_travel       job  
0              0.0  0.003829  
1              0.0  0.005747  
2              0.0  0.002843  
3              0.0  0.001081  
4              0.0  0.004741

6) Ahora nos queda pasar a variable numérica el Estado, son 51 distintos. De igual manera que hemos razonado anteriormente con las categorías de merchants, los Estados de Estados Unidos es una variable con "categorías fijas", en el sentido de que esta organización del país está establecida de esta manera por Leyes y un cambio en los Estados podría darse, pero después de un largo proceso de formalización. Hemos comprobado que en Train tenemos 51 valores distintos de Estados, **así que podemos asumir que no tendremos Estados nuevos en el conjunto de Test ni cuando el sistema entre en Producción**

Vamos a aplicar One-Hot Encoder:

In [23]:
# One-Hot-Encoding for State
ohe_state = preprocessing.OneHotEncoder(categories='auto')
state_encoded = ohe_state.fit_transform(df[["state"]])

states = ohe_state.get_feature_names_out()

# Concatenate sparse matrix with the Dataframe, droping "old" text-like category column
df = pd.concat([df.drop(columns=["state"]), pd.DataFrame(state_encoded.toarray(), columns = states)], axis=1)

In [24]:
df.head()

cc_num     amt  gender          city    zip      lat      long  \
0   213157767990030    4.39       0  Rock Springs  82901  41.6060 -109.2300   
1  3553629419254918   14.96       0        Conway  98238  48.3400 -122.3456   
2   376028110684021   32.43       1     Meadville  64659  39.7795  -93.3014   
3   180094419304907   66.36       0      Sun City  92585  33.7467 -117.1721   
4  3501942333500073  115.35       0       Phoenix  85086  33.8155 -112.1202   

   city_pop  merch_lat  merch_long  is_fraud  Year  Month  Unemployed percent  \
0     27971  41.971530 -109.512691         0  2019      2                 3.6   
1        85  49.088724 -123.344203         0  2020      3                 5.2   
2       964  39.155293  -92.421151         0  2019      7                 3.1   
3     54287  33.720708 -116.381735         0  2020      6                13.8   
4   1312922  34.542705 -111.282381         0  2019      5                 4.8   

   close_s&p_500  change_usd_eur  crime_rate  age  week_number  day_of_week  \
0        2709.80          0.8828       215.0   39            7            0   
1        2237.40          0.9349       293.7   39           12            6   
2        2975.95          0.8910       499.6   48           27            6   
3        3053.24          0.8907       442.0   51           26            5   
4        2850.96          0.8919       447.1   23           20            2   

   day  hour  is_holiday  category_entertainment  category_food_dining  \
0   11    14       False                     0.0                   0.0   
1   22    21       False                     0.0                   0.0   
2    7    10       False                     0.0                   0.0   
3   27    23       False                     0.0                   0.0   
4   15    12       False                     0.0                   0.0   

   category_gas_transport  category_grocery_net  category_grocery_pos  \
0                     0.0                   0.0                   0.0   
1                     0.0                   0.0                   0.0   
2                     0.0                   0.0                   0.0   
3                     0.0                   0.0                   0.0   
4                     0.0                   0.0                   0.0   

   category_health_fitness  category_home  category_kids_pets  \
0                      0.0            0.0                 0.0   
1                      0.0            0.0                 1.0   
2                      0.0            0.0                 0.0   
3                      0.0            1.0                 0.0   
4                      0.0            1.0                 0.0   

   category_misc_net  category_misc_pos  category_personal_care  \
0                0.0                0.0                     0.0   
1                0.0                0.0                     0.0   
2                0.0                1.0                     0.0   
3                0.0                0.0                     0.0   
4                0.0                0.0                     0.0   

   category_shopping_net  category_shopping_pos  category_travel       job  \
0                    1.0                    0.0              0.0  0.003829   
1                    0.0                    0.0              0.0  0.005747   
2                    0.0                    0.0              0.0  0.002843   
3                    0.0                    0.0              0.0  0.001081   
4                    0.0                    0.0              0.0  0.004741   

   state_AK  state_AL  state_AR  state_AZ  state_CA  state_CO  state_CT  \
0       0.0       0.0       0.0       0.0       0.0       0.0       0.0   
1       0.0       0.0       0.0       0.0       0.0       0.0       0.0   
2       0.0       0.0       0.0       0.0       0.0       0.0       0.0   
3       0.0       0.0       0.0       0.0       1.0       0.0       0.0   
4       0.0       0.0       0.0       1.0       0.0   

Probamos un `dummy_encoding` con City, porque no tenemos control sobre las ciudades que puedan aparecer en el futuro:

In [25]:
# Dummy encoding for city
df = pd.get_dummies(df, columns = ['city'])

In [26]:
df.head().T

0                 1  \
cc_num                          213157767990030  3553629419254918   
amt                                        4.39             14.96   
gender                                        0                 0   
zip                                       82901             98238   
lat                                      41.606             48.34   
long                                    -109.23         -122.3456   
city_pop                                  27971                85   
merch_lat                              41.97153         49.088724   
merch_long                          -109.512691       -123.344203   
is_fraud                                      0                 0   
Year                                       2019              2020   
Month                                         2                 3   
Unemployed percent                          3.6               5.2   
close_s&p_500                            2709.8            2237.4   
change_usd_eur                           0.8828            0.9349   
crime_rate                                215.0             293.7   
age                                          39                39   
week_number                                   7                12   
day_of_week                                   0                 6   
day                                          11                22   
hour                                         14                21   
is_holiday                                False             False   
category_entertainment                      0.0               0.0   
category_food_dining                        0.0               0.0   
category_gas_transport                      0.0               0.0   
category_grocery_net                        0.0               0.0   
category_grocery_pos                        0.0               0.0   
category_health_fitness                     0.0               0.0   
category_home                               0.0               0.0   
category_kids_pets                          0.0               1.0   
category_misc_net                           0.0               0.0   
category_misc_pos                           0.0               0.0   
category_personal_care                      0.0               0.0   
category_shopping_net                       1.0               0.0   
category_shopping_pos                       0.0               0.0   
category_travel                             0.0               0.0   
job                                    0.003829          0.005747   
state_AK                                    0.0               0.0   
state_AL                                    0.0               0.0   
state_AR                                    0.0               0.0   
state_AZ                                    0.0               0.0   
state_CA                                    0.0               0.0   
state_CO                                    0.0               0.0   
state_CT                                    0.0               0.0   
state_DC                                    0.0               0.0   
state_DE                                    0.0               0.0   
state_FL                                    0.0               0.0   
state_GA                                    0.0               0.0   
state_HI                                    0.0               0.0   
state_IA                                    0.0               0.0   
state_ID                                    0.0               0.0   
state_IL                                    0.0               0.0   
state_IN                                    0.0               0.0   
state_KS                                    0.0               0.0   
state_KY                                    0.0               0.0   
state_LA                                    0.0               0.0   
state_MA                                    0.0               0.0   
state_MD                                    

In [27]:
print(f'DataFrame Shape: ',df.shape)

DataFrame Shape:  (1481915, 994)


Como vemos, la dimensionalidad aumenta muchísimo al categorizar. Vamos a dropear la variable *city*, para no meter tanto ruido y porque consideramos que con la latitud y longitud, tenemos información más que suficiente respecto a la ubicación del propietario de la tarjeta:

In [28]:
city_enc = r'city_'

columns_to_drop = df.filter(regex=city_enc, axis=1).columns.difference(['city_pop']) # We want to keep city_pop
df.drop(columns=columns_to_drop, inplace=True)

Ya tenemos todas las variables como numéricas, vamos a ver si los formatos son correctos:

In [29]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1481915 entries, 0 to 1481914
Data columns (total 88 columns):
 #   Column                   Non-Null Count    Dtype  
---  ------                   --------------    -----  
 0   cc_num                   1481915 non-null  int64  
 1   amt                      1481915 non-null  float64
 2   gender                   1481915 non-null  int32  
 3   zip                      1481915 non-null  int64  
 4   lat                      1481915 non-null  float64
 5   long                     1481915 non-null  float64
 6   city_pop                 1481915 non-null  int64  
 7   merch_lat                1481915 non-null  float64
 8   merch_long               1481915 non-null  float64
 9   is_fraud                 1481915 non-null  int64  
 10  Year                     1481915 non-null  int64  
 11  Month                    1481915 non-null  int64  
 12  Unemployed percent       1481915 non-null  float64
 13  close_s&p_500            1481915 non-null 

In [30]:
print(f'DataFrame Shape: ',df.shape)

DataFrame Shape:  (1481915, 88)


7) Con la variable del código postal `zip` existe un caso particular: aparentemente se trata de una variable numérica, pero los modelos de ML no van a interpretar correctamente la información y se van a limitar a definir un "gradiente" según el valor del código postal, es decir, lo van a interpretar de forma cuantitativa. Por eso y porque tenemos información numérica real de las localizaciones, así como información útil de las ciudades como el número de habitantes, decidimos eliminar esta variable.

In [31]:
df = df.drop(['zip'], axis = 1)

In [32]:
df.head()

cc_num     amt  gender      lat      long  city_pop  merch_lat  \
0   213157767990030    4.39       0  41.6060 -109.2300     27971  41.971530   
1  3553629419254918   14.96       0  48.3400 -122.3456        85  49.088724   
2   376028110684021   32.43       1  39.7795  -93.3014       964  39.155293   
3   180094419304907   66.36       0  33.7467 -117.1721     54287  33.720708   
4  3501942333500073  115.35       0  33.8155 -112.1202   1312922  34.542705   

   merch_long  is_fraud  Year  Month  Unemployed percent  close_s&p_500  \
0 -109.512691         0  2019      2                 3.6        2709.80   
1 -123.344203         0  2020      3                 5.2        2237.40   
2  -92.421151         0  2019      7                 3.1        2975.95   
3 -116.381735         0  2020      6                13.8        3053.24   
4 -111.282381         0  2019      5                 4.8        2850.96   

   change_usd_eur  crime_rate  age  week_number  day_of_week  day  hour  \
0          0.8828       215.0   39            7            0   11    14   
1          0.9349       293.7   39           12            6   22    21   
2          0.8910       499.6   48           27            6    7    10   
3          0.8907       442.0   51           26            5   27    23   
4          0.8919       447.1   23           20            2   15    12   

   is_holiday  category_entertainment  category_food_dining  \
0       False                     0.0                   0.0   
1       False                     0.0                   0.0   
2       False                     0.0                   0.0   
3       False                     0.0                   0.0   
4       False                     0.0                   0.0   

   category_gas_transport  category_grocery_net  category_grocery_pos  \
0                     0.0                   0.0                   0.0   
1                     0.0                   0.0                   0.0   
2                     0.0                   0.0                   0.0   
3                     0.0                   0.0                   0.0   
4                     0.0                   0.0                   0.0   

   category_health_fitness  category_home  category_kids_pets  \
0                      0.0            0.0                 0.0   
1                      0.0            0.0                 1.0   
2                      0.0            0.0                 0.0   
3                      0.0            1.0                 0.0   
4                      0.0            1.0                 0.0   

   category_misc_net  category_misc_pos  category_personal_care  \
0                0.0                0.0                     0.0   
1                0.0                0.0                     0.0   
2                0.0                1.0                     0.0   
3                0.0                0.0                     0.0   
4                0.0                0.0                     0.0   

   category_shopping_net  category_shopping_pos  category_travel       job  \
0                    1.0                    0.0              0.0  0.003829   
1                    0.0                    0.0              0.0  0.005747   
2                    0.0                    0.0              0.0  0.002843   
3                    0.0                    0.0              0.0  0.001081   
4                    0.0                    0.0              0.0  0.004741   

   state_AK  state_AL  state_AR  state_AZ  state_CA  state_CO  state_CT  \
0       0.0       0.0       0.0       0.0       0.0       0.0       0.0   
1       0.0       0.0       0.0       0.0       0.0       0.0       0.0   
2       0.0       0.0       0.0       0.0       0.0       0.0       0.0   
3       0.0       0.0       0.0       0.0       1.0       0.0       0.0   
4       0.0       0.0       0.0       1.0       0.0       0.0       0.0   

   state_DC  state_DE  state_FL  state_GA  state_HI  state_IA  state_ID  \
0       0.0       0.0       0

8) Con la variable del número de tarjeta asociado a la transacción, `cc_num` nos pasa lo mismo que con `zip`: es un número, cantidad que no tiene correspondencia con una realidad cuantitativa real, se trata al fin y al cabo de un dato personal, hay que considerarlo igual que el nombre o el apellido. Por todo esto, la eliminamos:

In [33]:
df = df.drop(['cc_num'], axis = 1)

In [34]:
df.head()

amt  gender      lat      long  city_pop  merch_lat  merch_long  \
0    4.39       0  41.6060 -109.2300     27971  41.971530 -109.512691   
1   14.96       0  48.3400 -122.3456        85  49.088724 -123.344203   
2   32.43       1  39.7795  -93.3014       964  39.155293  -92.421151   
3   66.36       0  33.7467 -117.1721     54287  33.720708 -116.381735   
4  115.35       0  33.8155 -112.1202   1312922  34.542705 -111.282381   

   is_fraud  Year  Month  Unemployed percent  close_s&p_500  change_usd_eur  \
0         0  2019      2                 3.6        2709.80          0.8828   
1         0  2020      3                 5.2        2237.40          0.9349   
2         0  2019      7                 3.1        2975.95          0.8910   
3         0  2020      6                13.8        3053.24          0.8907   
4         0  2019      5                 4.8        2850.96          0.8919   

   crime_rate  age  week_number  day_of_week  day  hour  is_holiday  \
0       215.0   39            7            0   11    14       False   
1       293.7   39           12            6   22    21       False   
2       499.6   48           27            6    7    10       False   
3       442.0   51           26            5   27    23       False   
4       447.1   23           20            2   15    12       False   

   category_entertainment  category_food_dining  category_gas_transport  \
0                     0.0                   0.0                     0.0   
1                     0.0                   0.0                     0.0   
2                     0.0                   0.0                     0.0   
3                     0.0                   0.0                     0.0   
4                     0.0                   0.0                     0.0   

   category_grocery_net  category_grocery_pos  category_health_fitness  \
0                   0.0                   0.0                      0.0   
1                   0.0                   0.0                      0.0   
2                   0.0                   0.0                      0.0   
3                   0.0                   0.0                      0.0   
4                   0.0                   0.0                      0.0   

   category_home  category_kids_pets  category_misc_net  category_misc_pos  \
0            0.0                 0.0                0.0                0.0   
1            0.0                 1.0                0.0                0.0   
2            0.0                 0.0                0.0                1.0   
3            1.0                 0.0                0.0                0.0   
4            1.0                 0.0                0.0                0.0   

   category_personal_care  category_shopping_net  category_shopping_pos  \
0                     0.0                    1.0                    0.0   
1                     0.0                    0.0                    0.0   
2                     0.0                    0.0                    0.0   
3                     0.0                    0.0                    0.0   
4                     0.0                    0.0                    0.0   

   category_travel       job  state_AK  state_AL  state_AR  state_AZ  \
0              0.0  0.003829       0.0       0.0       0.0       0.0   
1              0.0  0.005747       0.0       0.0       0.0       0.0   
2              0.0  0.002843       0.0       0.0       0.0       0.0   
3              0.0  0.001081       0.0       0.0       0.0       0.0   
4              0.0  0.004741       0.0       0.0       0.0       1.0   

   state_CA  state_CO  state_CT  state_DC  state_DE  state_FL  state_GA  \
0       0.0       0.0       0.0       0.0       0.0       0.0       0.0   
1       0.0       0.0       0.0       0.0       0.0       0.0       0.0   
2       0.0       0.0       0.0       0.0       0.0       0.0       0.0   
3       1.0       0.0       0.0       0.0       0.0       0.0       0.0   
4       0.0       0.0       0.0       0.0

9) Cambiamos a valor numérico también la variable booleana `is_holiday`, por si acaso los modelos no la permiten:

In [35]:
df['is_holiday'] = df['is_holiday'].astype(int)

In [36]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1481915 entries, 0 to 1481914
Data columns (total 86 columns):
 #   Column                   Non-Null Count    Dtype  
---  ------                   --------------    -----  
 0   amt                      1481915 non-null  float64
 1   gender                   1481915 non-null  int32  
 2   lat                      1481915 non-null  float64
 3   long                     1481915 non-null  float64
 4   city_pop                 1481915 non-null  int64  
 5   merch_lat                1481915 non-null  float64
 6   merch_long               1481915 non-null  float64
 7   is_fraud                 1481915 non-null  int64  
 8   Year                     1481915 non-null  int64  
 9   Month                    1481915 non-null  int64  
 10  Unemployed percent       1481915 non-null  float64
 11  close_s&p_500            1481915 non-null  float64
 12  change_usd_eur           1481915 non-null  float64
 13  crime_rate               1481915 non-null 

## Creación de variables sintéticas

Hemos visto que tenemos la latitud y longitud del domicilio del propietario y la del comercio en donde se hizo la compra con tarjeta, un dato muy interesante es el de la **distancia entre un punto y el otro**, ya que normalmente hacemos compras cercanas a nuestro domicilio. Además, esto nos permite añadir más información al modelo respecto a las ubicaciones y evitar que se "fije" en `lat` y `merch_lat` de forma errónea, ya que matemáticamente correlan mucho según vimos en la Matriz de Correlación.

Por otro lado, hay que tener en mente que se puede tratar de compras online, claro.

Existe una librería `geopy` que facilita esta tarea https://pypi.org/project/geopy/

*geopy is a Python client for several popular geocoding web services.*

*geopy makes it easy for Python developers to locate the coordinates of addresses, cities, countries, and landmarks across the globe using third-party geocoders and other data sources.*

*geopy includes geocoder classes for the OpenStreetMap Nominatim, Google Geocoding API (V3), and many other geocoding services. The full list is available on the Geocoders doc section. Geocoder classes are located in geopy.geocoders.*

*geopy is tested against CPython (versions 3.7, 3.8, 3.9, 3.10, 3.11, 3.12) and PyPy3. geopy 1.x line also supported CPython 2.7, 3.4 and PyPy2.*

*© geopy contributors 2006-2018 (see AUTHORS) under the MIT License.*

In [37]:
!pip install geopy

Probamos qué tal funciona:

In [38]:
from geopy.distance import geodesic

origin = (30.172705, 31.526725)

dist = (30.288281, 31.732326)

print(geodesic(origin, dist).miles)

14.64994773134371


Creamos una nueva variable llamada `merch_home_distance` y la calculamos con la función `geodesic` de Geopy

In [39]:
#df['merch_home_distance'] = df.apply(lambda x : geodesic(df['lat']))

df['merch_home_distance'] = df.apply(lambda row: geodesic((row['lat'], row['long']), 
                                                          (row['merch_lat'], row['merch_long'])).miles, axis=1) #In miles

In [40]:
df.head()

amt  gender      lat      long  city_pop  merch_lat  merch_long  \
0    4.39       0  41.6060 -109.2300     27971  41.971530 -109.512691   
1   14.96       0  48.3400 -122.3456        85  49.088724 -123.344203   
2   32.43       1  39.7795  -93.3014       964  39.155293  -92.421151   
3   66.36       0  33.7467 -117.1721     54287  33.720708 -116.381735   
4  115.35       0  33.8155 -112.1202   1312922  34.542705 -111.282381   

   is_fraud  Year  Month  Unemployed percent  close_s&p_500  change_usd_eur  \
0         0  2019      2                 3.6        2709.80          0.8828   
1         0  2020      3                 5.2        2237.40          0.9349   
2         0  2019      7                 3.1        2975.95          0.8910   
3         0  2020      6                13.8        3053.24          0.8907   
4         0  2019      5                 4.8        2850.96          0.8919   

   crime_rate  age  week_number  day_of_week  day  hour  is_holiday  \
0       215.0   39            7            0   11    14           0   
1       293.7   39           12            6   22    21           0   
2       499.6   48           27            6    7    10           0   
3       442.0   51           26            5   27    23           0   
4       447.1   23           20            2   15    12           0   

   category_entertainment  category_food_dining  category_gas_transport  \
0                     0.0                   0.0                     0.0   
1                     0.0                   0.0                     0.0   
2                     0.0                   0.0                     0.0   
3                     0.0                   0.0                     0.0   
4                     0.0                   0.0                     0.0   

   category_grocery_net  category_grocery_pos  category_health_fitness  \
0                   0.0                   0.0                      0.0   
1                   0.0                   0.0                      0.0   
2                   0.0                   0.0                      0.0   
3                   0.0                   0.0                      0.0   
4                   0.0                   0.0                      0.0   

   category_home  category_kids_pets  category_misc_net  category_misc_pos  \
0            0.0                 0.0                0.0                0.0   
1            0.0                 1.0                0.0                0.0   
2            0.0                 0.0                0.0                1.0   
3            1.0                 0.0                0.0                0.0   
4            1.0                 0.0                0.0                0.0   

   category_personal_care  category_shopping_net  category_shopping_pos  \
0                     0.0                    1.0                    0.0   
1                     0.0                    0.0                    0.0   
2                     0.0                    0.0                    0.0   
3                     0.0                    0.0                    0.0   
4                     0.0                    0.0                    0.0   

   category_travel       job  state_AK  state_AL  state_AR  state_AZ  \
0              0.0  0.003829       0.0       0.0       0.0       0.0   
1              0.0  0.005747       0.0       0.0       0.0       0.0   
2              0.0  0.002843       0.0       0.0       0.0       0.0   
3              0.0  0.001081       0.0       0.0       0.0       0.0   
4              0.0  0.004741       0.0       0.0       0.0       1.0   

   state_CA  state_CO  state_CT  state_DC  state_DE  state_FL  state_GA  \
0       0.0       0.0       0.0       0.0       0.0       0.0       0.0   
1       0.0       0.0       0.0       0.0       0.0       0.0       0.0   
2       0.0       0.0       0.0       0.0       0.0       0.0       0.0   
3       1.0       0.0       0.0       0.0       0.0       0.0       0.0   
4       0.0       0.0       0.0       0.0

Las categorías con la palabra clave `_net` son de compras realizadas por Internet (*category_grocery_net , category_shopping_net, category_misc_net*), en estos casos la distancia relativa entre el domicilio del propietario de la tarjeta y la ubicación del comercio no tiene sentido. Vamos a crear una variable que indique esto:

In [41]:
# Identify columns ending with '_net'
online_columns = [col for col in df.columns if col.endswith('_net')]

# Create a new column indicating if it's an online shopping category
df['is_online_shopping'] = df[online_columns].apply(lambda row: 1 if row.any() else 0, axis=1)

In [42]:
df.head()

amt  gender      lat      long  city_pop  merch_lat  merch_long  \
0    4.39       0  41.6060 -109.2300     27971  41.971530 -109.512691   
1   14.96       0  48.3400 -122.3456        85  49.088724 -123.344203   
2   32.43       1  39.7795  -93.3014       964  39.155293  -92.421151   
3   66.36       0  33.7467 -117.1721     54287  33.720708 -116.381735   
4  115.35       0  33.8155 -112.1202   1312922  34.542705 -111.282381   

   is_fraud  Year  Month  Unemployed percent  close_s&p_500  change_usd_eur  \
0         0  2019      2                 3.6        2709.80          0.8828   
1         0  2020      3                 5.2        2237.40          0.9349   
2         0  2019      7                 3.1        2975.95          0.8910   
3         0  2020      6                13.8        3053.24          0.8907   
4         0  2019      5                 4.8        2850.96          0.8919   

   crime_rate  age  week_number  day_of_week  day  hour  is_holiday  \
0       215.0   39            7            0   11    14           0   
1       293.7   39           12            6   22    21           0   
2       499.6   48           27            6    7    10           0   
3       442.0   51           26            5   27    23           0   
4       447.1   23           20            2   15    12           0   

   category_entertainment  category_food_dining  category_gas_transport  \
0                     0.0                   0.0                     0.0   
1                     0.0                   0.0                     0.0   
2                     0.0                   0.0                     0.0   
3                     0.0                   0.0                     0.0   
4                     0.0                   0.0                     0.0   

   category_grocery_net  category_grocery_pos  category_health_fitness  \
0                   0.0                   0.0                      0.0   
1                   0.0                   0.0                      0.0   
2                   0.0                   0.0                      0.0   
3                   0.0                   0.0                      0.0   
4                   0.0                   0.0                      0.0   

   category_home  category_kids_pets  category_misc_net  category_misc_pos  \
0            0.0                 0.0                0.0                0.0   
1            0.0                 1.0                0.0                0.0   
2            0.0                 0.0                0.0                1.0   
3            1.0                 0.0                0.0                0.0   
4            1.0                 0.0                0.0                0.0   

   category_personal_care  category_shopping_net  category_shopping_pos  \
0                     0.0                    1.0                    0.0   
1                     0.0                    0.0                    0.0   
2                     0.0                    0.0                    0.0   
3                     0.0                    0.0                    0.0   
4                     0.0                    0.0                    0.0   

   category_travel       job  state_AK  state_AL  state_AR  state_AZ  \
0              0.0  0.003829       0.0       0.0       0.0       0.0   
1              0.0  0.005747       0.0       0.0       0.0       0.0   
2              0.0  0.002843       0.0       0.0       0.0       0.0   
3              0.0  0.001081       0.0       0.0       0.0       0.0   
4              0.0  0.004741       0.0       0.0       0.0       1.0   

   state_CA  state_CO  state_CT  state_DC  state_DE  state_FL  state_GA  \
0       0.0       0.0       0.0       0.0       0.0       0.0       0.0   
1       0.0       0.0       0.0       0.0       0.0       0.0       0.0   
2       0.0       0.0       0.0       0.0       0.0       0.0       0.0   
3       1.0       0.0       0.0       0.0       0.0       0.0       0.0   
4       0.0       0.0       0.0       0.0

## Guardamos el conjunto de Train procesado, con todas las variables en formato numérico:

In [43]:
# From Jupyter:
df.to_csv('./data/train_processed.csv', index=False)

## Creamos nuestra función de Preprocesado para aplicarla en el conjunto de Test:

In [44]:
from sklearn import preprocessing
# !pip install geopy
from geopy.distance import geodesic

La función `dataset_processing( )` recoge todos los pasos que hemos hecho sobre el conjunto de Train para pasar a dato numérico, añadir información adicional, etc. En el caso de los Encoders, tomaremos directamente el encoder definido sobre los datos de Train y haremos solamente un `transform`, para no inferir sobre datos de Test.

In [45]:
def dataset_processing (df):
    
    # Dropping variables
    df = df.drop(['state_abbr','trans_date_trans_time', 'time', 'date','zip','city', 'cc_num'], axis = 1)
    
    # Weekdays into nums
    df['day_of_week'] = df['day_of_week'].map({
    'Monday': 0,
    'Tuesday': 1,
    'Wednesday': 2,
    'Thursday': 3,
    'Friday': 4,
    'Saturday': 5,
    'Sunday': 6 })
    
    # is_holiday as num type
    df['is_holiday'] = df['is_holiday'].astype(int)
    
    # Label Encoder **previously defined in Train** for Gender
    df["gender"] = le_gender.transform(df["gender"])
    
    # One Hot encoder **previously defined in Train** for merchants categories
    cat_encoded = ohe_category.transform(df[["category"]])
    #merch_cats = ohe_category.get_feature_names_out()
    df = pd.concat([df.drop(columns=["category"]), pd.DataFrame(cat_encoded.toarray(),columns=merch_cats)], axis=1)
    
    # Target encoder **previously defined in Train** for jobs
    job_encoded = targ_job.transform(df[['job']])
    #te_job = targ_job.feature_names_in_
    df = pd.concat([df.drop(columns=["job"]), pd.DataFrame(job_encoded, columns=te_job)], axis=1)
    
    # One-Hot-Encoding **previously defined in Train** for State
    state_encoded = ohe_state.transform(df[["state"]])
    #states = ohe_state.get_feature_names_out()
    df = pd.concat([df.drop(columns=["state"]), pd.DataFrame(state_encoded.toarray(), columns = states)], axis=1)
    
    # Distance
    df['merch_home_distance'] = df.apply(lambda row: geodesic((row['lat'], row['long']), 
                                                          (row['merch_lat'], row['merch_long'])).miles, axis=1) #In miles
    # Online shopping transactions
    online_columns = [col for col in df.columns if col.endswith('_net')]
    df['is_online_shopping'] = df[online_columns].apply(lambda row: 1 if row.any() else 0, axis=1)
    
    return df

# Cargamos csv de Test:

Aplicamos el pre-procesamiento necesario para unificar formatos con respecto a Train.


**NOTA: Todos los pasos que vamos a hacer se basan en los datos e información del conjunto de Train, de Test no tenemos información alguna. Por eso, principalmente, hemos creado la función de preprocesado**

Cargamos el csv con los datos de Test:

In [46]:
df_test = pd.read_csv('./data/test.csv', sep=';', decimal='.')

Aplicamos preprocesamiento:

In [47]:
test_processed = dataset_processing (df_test)

In [48]:
test_processed.shape

(370479, 88)

Guardamos el conjunto de Test procesado:

In [49]:
# From Jupyter:
test_processed.to_csv('./data/test_processed.csv', index=False)

Y, por último, también guardamos las variables de los encoders para poder usarlos más adelante:

In [50]:
%store le_gender
%store ohe_category
%store targ_job
%store ohe_state

Stored 'le_gender' (LabelEncoder)
Stored 'ohe_category' (OneHotEncoder)
Stored 'targ_job' (TargetEncoder)
Stored 'ohe_state' (OneHotEncoder)
